synapseutils/migrate_functions.py

```diff
@@ -0,0 +1,736 @@
+        try:
+            migrated_version, from_file_handle_id, to_file_handle_id = _migrate_file_version(
+                syn,
+                entity,
+                storage_location_id,
+                version,
+            )
+            mapping[migrated_version] = (from_file_handle_id, to_file_handle_id)
+
+        except _AlreadyMigratedException:
+            # specified revision is already stored in the destination storage location
+            pass
+
+    return mapping
+
+
+def _create_new_file_version(syn, entity, storage_location_id):
+    existing_file_handle_id = entity.dataFileHandleId
+    existing_file_name = entity._file_handle['fileName']
+    existing_storage_location_id = entity._file_handle['storageLocationId']
+
+    if str(existing_storage_location_id) == str(storage_location_id):
+        logging.info(
+            'Skipped creating a new version of file %s, it is already in the destination storage location (%s)',
+            entity.id,
+            storage_location_id
+        )
+        raise _AlreadyMigratedException()
+
+    source_file_handle_association = {
+        'fileHandleId': existing_file_handle_id,
+        'associateObjectId': entity.id,
+        'associateObjectType': 'FileEntity',
+    }
+
+    new_file_handle_id = multipart_copy(
+        syn,
+        source_file_handle_association,
+        dest_file_name=existing_file_name,
+        storage_location_id=storage_location_id,
+    )
+
+    entity.dataFileHandleId = new_file_handle_id
+    entity = syn.store(entity)
+
+    return entity.versionNumber, existing_file_handle_id, new_file_handle_id
+
+
+def _migrate_all_file_versions(syn, entity, storage_location_id):
+    mapping = {}
+    entity_id = utils.id_of(entity)
+
+    for version in syn._GET_paginated("/entity/{id}/version".format(id=entity_id)):
+        version_number = version['versionNumber']
+        try:
+            _, from_file_handle_id, to_file_handle_id = _migrate_file_version(
+                syn,
+                entity,
+                storage_location_id,
+                version_number,
+            )
+
+            mapping[version_number] = (from_file_handle_id, to_file_handle_id)
+
+        except _AlreadyMigratedException:
+            pass
+
+    return mapping
+
+
+def _migrate_file_version(syn, entity, storage_location_id, version):
+    # if walking a container the children entities passed will be the light weight
+    # dictionary representation of the FileEntity, so we retrieve the full entity
+    entity_id = utils.id_of(entity)
+    entity = syn.get(entity, downloadFile=False, version=version)
+    version = entity.versionNumber
+
+    existing_file_handle_id = entity.dataFileHandleId
+    existing_file_name = entity._file_handle['fileName']
+    existing_storage_location_id = entity._file_handle['storageLocationId']
+
+    if str(existing_storage_location_id) == str(storage_location_id):
+        logging.info(
```

Comment:
warn?

Reply:
Removed this logging statement, already migrated items are recorded in the index.